### PR TITLE
JAMES-3121 Cassandra tables tunings

### PR DESCRIPTION
**Pull request by blackheaven**

### CODE REVIEW

File: upgrade-instructions.md [hunk 1]
```diff
@@ -69,6 +69,28 @@ we regenerate invalid UidValidity upon reads.
...
+#### Upgrade procedure
+
+Alter `deletedMailsV2` and `enqueuedMailsV3` with this `cqlsh` commands:
+
+```
+ALTER TABLE james_keyspace.deletedMailsV2 WITH compaction = { 'class' :  'TimeWindowCompactionStrategy'  };
+ALTER TABLE james_keyspace.enqueuedMailsV3 WITH compaction = { 'class' :  'TimeWindowCompactionStrategy'  };
+ALTER TABLE james_keyspace.deletedMailsV2 WITH read_repair_chance = 0.0;
+ALTER TABLE james_keyspace.enqueuedMailsV3 WITH read_repair_chance = 0.0;
+```
```
Comment (chibenwa) [inline upgrade-instructions.md]: Explain that compaction  strategies changes requires a compaction to actually take effect.

File: upgrade-instructions.md [hunk 2]
```diff
@@ -69,6 +69,28 @@ we regenerate invalid UidValidity upon reads.
...
+SHA-1 a6613bc91b c8792c64bf
+
+JIRA: https://issues.apache.org/jira/browse/JAMES-3121
+
+Our usage of Cassandra for the time series have been improved by fine tuning the compaction strategy and
+the read repair option.
+
+#### Upgrade procedure
+
+Alter `deletedMailsV2` and `enqueuedMailsV3` with this `cqlsh` commands:
```
Comment (chibenwa) [inline upgrade-instructions.md]: In order to unlock these improvements we advise you running the following commands on existing instalations

File: upgrade-instructions.md [hunk 3]
```diff
@@ -69,6 +69,28 @@ we regenerate invalid UidValidity upon reads.
...
  resynchronisation. (one chance out of two billions).
 
+### Tune Cassandra time serie tables options
+
+Date 18/03/2020
+
+SHA-1 a6613bc91b c8792c64bf
+
+JIRA: https://issues.apache.org/jira/browse/JAMES-3121
+
```
Comment (chibenwa) [inline upgrade-instructions.md]: Impacted product: Guice distributed James

File: server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/CassandraMailQueueViewModule.java [hunk 4]
```diff
@@ -84,7 +87,9 @@
             " when a mail is dequeued from a mail queue, the record associated with that mail still available in this" +
             " table and will not be deleted immediately regarding to the performance impacts," +
             " but after some scheduled tasks")
-        .options(options -> options)
+        .options(options -> options
+            .compactionOptions(SchemaBuilder.timeWindowCompactionStrategy())
```
Comment (chibenwa) [inline server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/CassandraMailQueueViewModule.java]: The time window looks like an important configuration parameter:

```
AND compaction = {'compaction_window_size': '1', 
    				  'compaction_window_unit': 'MINUTES', 
    				  'class': 'org.apache.cassandra.db.compaction.TimeWindowCompacti
```

File: upgrade-instructions.md (hunk 1 above)
Comment (blackheaven) [reply upgrade-instructions.md]: @remk find a document explaining that it automatically triggers it

File: server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/CassandraMailQueueViewModule.java (hunk 4 above)
Comment (blackheaven) [reply server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/CassandraMailQueueViewModule.java]: 1 minute seems too often

File: upgrade-instructions.md (hunk 1 above)
Comment (chibenwa) [reply upgrade-instructions.md]: What do you mean?

Why does it makes sense here?

**Comment by blackheaven** [comment]
rebased for comment edition and order

File: server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/CassandraMailQueueViewModule.java (hunk 4 above)
Comment (chibenwa) [reply server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/CassandraMailQueueViewModule.java]: How about 1 hour (the default window size used for primary keys?)

File: upgrade-instructions.md (hunk 1 above)
Comment (remk) [reply upgrade-instructions.md]: https://support.datastax.com/hc/en-us/articles/213370546-Change-CompactionStrategy-and-sub-properties-via-JMX

Doing via cqlsh with alter table trigger a compaction automatically on all nodes which can be on the heavy side, there is recommended alternatives using JMX but for this tables which should not grow too much the cqlsh approach seems good. There is no reference to having to trigger a compaction with nodetool in the documentation.
https://docs.datastax.com/en/cql-oss/3.3/cql/cql_reference/cqlAlterTable.html#cqlAlterTable__alter-compression

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/modules/CassandraApplicableFlagsModule.java [hunk 5]
```diff
@@ -35,7 +35,7 @@
         .comment("Holds flags being used on specific mailboxes. As system flags are implicit, this table " +
             "stores user flags.")
         .options(options -> options
-            .compactionOptions(SchemaBuilder.leveledStrategy())
+            .compactionOptions(SchemaBuilder.sizedTieredStategy())
```
Comment (remk) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/modules/CassandraApplicableFlagsModule.java]: what was your reasoning to move from LCS to STS ?
Can you explain it in the commit description.

https://docs.datastax.com/en/dse/5.1/dse-dev/datastax_enterprise/config/configChooseCompactStrategy.html

**Comment by blackheaven** [comment]
rebased for comments

**Comment by blackheaven** [comment]
rebased for doc conflicts

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**